sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEPropagateEmptyRelation.scala

```diff
@@ -53,8 +54,8 @@ object AQEPropagateEmptyRelation extends PropagateEmptyRelationBase {
       empty(j)
   }
 
-  // TODO we need use transformUpWithPruning instead of transformUp
-  def apply(plan: LogicalPlan): LogicalPlan = plan.transformUp {
+  def apply(plan: LogicalPlan): LogicalPlan = plan.transformUpWithPruning(
+    _.containsAnyPattern(LOGICAL_QUERY_STAGE, LOCAL_RELATION, TRUE_OR_FALSE_LITERAL), ruleId) {
```

Comment:
Could you add a comment to explain where `LOCAL_RELATION` and `TRUE_OR_FALSE_LITERAL` are matched?

Reply:
Added, but not sure that's you want. please tell me if you need anything else.

Reply:
Let's add the comment at line 61. 
I always have to check the base class for the patterns on reading the code. So having the comments makes future development easier.
Let's document LOGICAL_QUERY_STAGE as well.

Reply:
ok, done